lib/elixir/lib/calendar.ex

```diff
@@ -0,0 +1,87 @@
+defmodule Calendar do
+  @moduledoc """
+  This module defines the responsibilities for working with
+  calendars, dates, times and datetimes in Elixir.
+
+  Currently it defines types but may define the calendar
+  behaviour in future versions. For the actual date, time
+  and datetime structures, see `Date`, `Time`, `NaiveDateTime`
+  and `DateTime`.
+
+  Note the year, month, day, etc designations are over specified
+  (i.e. an integer instead of 1..12 for months) because different
+  calendars may have a different number of numbers and so on.
+  """
+
+  @type year        :: integer
+  @type month       :: integer
+  @type day         :: integer
+  @type hour        :: integer
+  @type minute      :: integer
+  @type second      :: integer
+  @type microsecond :: integer
+
+  @typedoc "A calendar implementation"
+  @type calendar    :: module
+
+  @typedoc "The timezone designation (never an abbreviation)"
+  @type time_zone   :: String.t
+
+  @typedoc "The timezone UTC offset in seconds"
+  @type utc_offset  :: integer
+
+  @typedoc "The timezone standard offset in seconds (not zero in summer times)"
+  @type std_offset  :: integer
+end
+
+defmodule Date do
+  @moduledoc """
+  A date implementation.
+  """
+  defstruct year: 1970, month: 1, day: 1
+  @type t :: %__MODULE__{year: Calendar.year, month: Calendar.month, day: Calendar.day}
+end
+
+defmodule Time do
+  @moduledoc """
+  A time implementation.
+  """
+  defstruct hour: 0, minute: 0, second: 0, microsecond: 0
+  @type t :: %__MODULE__{hour: Calendar.hour, minute: Calendar.minute,
+                         second: Calendar.second, microsecond: Calendar.microsecond}
+end
+
+defmodule NaiveDateTime do
+  @moduledoc """
+  A naive datetime implementation (without a time zone).
+
+  The naive bit implies this datetime representation does
+  not have a timezone. This means the datetime may not
+  actually exist in certain areas in the world even though
+  it is valid.
+
+  For example, when daylight saving changes are adopted
+  by a region, the clock typically moves forward or backward
+  by one hour. This means certain datetimes never occur or
+  may occur more than once.
+  """
+  defstruct year: 1970, month: 1, day: 1, hour: 0, minute: 0, second: 0,
+            microsecond: 0, calendar: Calendar.ISO
+  @type t :: %__MODULE__{year: Calendar.year, month: Calendar.month, day: Calendar.day,
+                         calendar: Calendar.calendar, hour: Calendar.hour, minute: Calendar.minute,
+                         second: Calendar.second, microsecond: Calendar.microsecond}
+end
+
+defmodule DateTime do
+  @moduledoc """
+  A datetime implementation with a time zone.
+  """
+  defstruct year: 1970, month: 1, day: 1, hour: 0, minute: 0, second: 0,
+            microsecond: 0, calendar: Calendar.ISO, time_zone: nil,
```

Comment:
Why not have all default values except `calendar: Calendar.ISO` be nil? For sure there will be people who will construct a struct and e.g. not set the year. If that happens let's look at the two situations:

1) No default: Eventually an error happens. Let it crash. Great.
2) Default 1970 year: The default silently paves over the input error and some function happily does some expensive million dollar transaction based on the year 1970. Sad face. (I talked about this for about half a minute here at around 22:50 here https://www.youtube.com/watch?v=keUbVvMJeKY&t=22m50s )